### DISCUSSION TITLE
Xfrmi routing not working

### DISCUSSION THREAD

**Discussion by robdyck** [strongswan/strongswan#1871] [General]
The scenario is a road warrior using a laptop tethered to a phone. xfrmi ipsec0 was added and up. In swanctl.conf the child is configured to add routing and it references the ID of the xfrmi.

Initiation goes well. Virtual IP's are added as well as custom DNS. Iptables accepts anything.
Pings fail.

The configuration is identical to a working policy route except for the addition of the xfrmi.
I looked at the routing table 220.
```
default dev ipsec0 proto static src 192.168.1.70
```
and compared it to the routing table from a policy example.
```
default via 192.168.42.129 dev enp0s20u3 proto static src 192.168.1.76
```
I wondered if it was a next-hop problem. I deleted the route and added something similar to the policy route.
```
ip r add default via 192.168.42.129 dev ipsec0  proto static src 192.168.1.70 t 220

Error: Nexthop has invalid gateway.
```
The addition didn't work but after some searching I found the "onlink" keyword.
```
ip r add default via 192.168.42.129 dev ipsec0  onlink proto static src 192.168.1.70 t 220
```

It still wasn't working. Also I cannot terminate the SA.
Any suggestions.

**Comment by tobiasbrunner** [comment]
> I looked at the routing table 220

Looks fine, but please be aware that directing the default route (or any other route that covers the IKE peer's IP) via an XFRM interface will also affect IKE and ESP traffic if you don't take appropriate measures. See the [notes in the docs](https://docs.strongswan.org/docs/5.9/features/routeBasedVpn.html#_configuration_2), which are actually incomplete. You also need to set the same mark in `set_mark_out` in swanctl.conf so that ESP traffic is also avoiding the routing table.

> Also I cannot terminate the SA.

What do you mean?

**Comment by robdyck** [comment]
By terminate I mean swanctl -t -i xfrmi.

I have compared a working configuration without xfrmi and non-working configuration with xfrmi

Road warrior with VIP using laptop tethered to phone. Tethering creates interface enp0s20u3.

Working
outgoing policy - src 192.168.1.77 dst 0.0.0.0, templ src 192.168.42.184 dst server public address
route 220 - default via 192.168.42.129 dev enp0s20u3 proto static src 192.168.1.77
traffic - interface enp0s20u3 src 192.168.42.184 dst server public address encrypted

Not working using xfrmi ipsec0
outgoing policy - src 192.168.1.79 dst 0.0.0.0 templ src 192.168.42.184 dst server public address
route 220 - default dev ipsec0 proto static src 192.168.1.79
traffic - interface enp0s20u3 src 192.168.42.184 dst 192.168.42.129 not encrypted

**Reply by tobiasbrunner** [reply]
> By terminate I mean swanctl -t -i xfrmi.

But what doesn't work?




> traffic - interface enp0s20u3 src 192.168.42.184 dst 192.168.42.129 not encrypted

So check the other routes and routing rules. Maybe the one via that interface has higher priority.

**Comment by robdyck** [comment]
Outgoing packets seem to be falling into a black hole. There are no responses ( except DNS which is going to the phone ) . I admit that my understanding of routing is limited. However I am not trying anything exotic. The main route is set by the system when the laptop is tethered to the phone. I chose to let the system setup the route for the SA. So the only routes are ones set by the system. My understanding is that rule 220 applies to all outgoing packets and if table 220 exists, main is not used.

I followed the example given except for the traffic selectors. The idea is that the road warrior is just another host on the LAN. I get that using xfrmi can interfere with IKE but I am not getting that far.

Am I understanding this. The xfrmi is created and given an ID. This ID is referenced in the child configuration. The daemon creates the policy based on the traffic selectors and associates that policy with xfrmi. The daemon creates a route which directs all outgoing traffic to the xfrmi. The underlying interface has no bearing on this unless an interface is referenced in the child configuration, which it's not.

Outgoing is definitely using table 220
```
ip r get 8.8.8.8
8.8.8.8 via 192.168.42.129 dev ipsec0 table 220 src 192.168.1.79 uid 0
ip r s t 220
default via 192.168.42.129 dev ipsec0 proto static src 192.168.1.79
```
I noticed something interesting in wireshark on the capture setup page. It graphs the traffic on the interfaces. The graphs for ipsec0 and enp0s20u3 seem identical. enp0s20u3  shows as proto ether and ipsec0 shows as proto raw. So there is traffic on ipsec0 but it doesn't show in the capture.

The system logs are also interesting. After the SA is up there are repeated logs.
ipsec0: local routing loop detected

I have taken the precaution of deleting ipsec0 and re-adding it making sure the ID corresponds with the IDs referenced in the child configuration.

**Reply by tobiasbrunner** [reply]
I already told you what you have to configure to avoid the routing loop.
